### ISSUE TITLE
ExpiringMap:remove(Object key) throws java.util.NoSuchElementException 

### ISSUE THREAD

**Issue by traianmatei** [jhalterman/expiringmap#29]
Under heavy load with multiple concurrent threads performing add/remove this error is thrown:

> java.util.NoSuchElementException
		at java.util.LinkedHashMap$LinkedHashIterator.nextEntry(LinkedHashMap.java:396)
		at java.util.LinkedHashMap$ValueIterator.next(LinkedHashMap.java:409)
		at net.jodah.expiringmap.ExpiringMap$EntryLinkedHashMap.first(ExpiringMap.java:266)
		at net.jodah.expiringmap.ExpiringMap.remove(ExpiringMap.java:838)

The reason for it: the WriteLock is released to early

```
public V remove(Object key) {
    Assert.notNull(key, "key");
    ExpiringEntry<K, V> entry = null;
    writeLock.lock();
    try {
      entry = entries.remove(key);
    } finally {//This finally block should be moved to the end of the method
      writeLock.unlock();
    }
    if (entry == null)
      return null;
    if (entry.cancel(false))
      scheduleEntry(entries.first());
    return entry.getValue();
  }
```
BTW the method  _public boolean remove(Object key, Object value)_ is correctly implemented.

**Comment by jhalterman** [comment]
Thanks a lot for reporting this. Will try to reproduce later today...

**Comment by jhalterman** [comment]
Released 0.5.5 with this fix.

**Comment by traianmatei** [comment]
Thanks a lot :)
